contracts/access_control/lib.rs

```diff
@@ -0,0 +1,245 @@
+        /// Creates a new contract.
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            // This call is required in order to correctly initialize the
+            // `Mapping`s of our contract.
+            ink_lang::utils::initialize_contract(|contract| Self::new_init(contract))
+        }
+
+        /// Initializes the contract.
+        ///
+        /// caller is granted admin and owner piviledges
+        fn new_init(&mut self) {
+            let caller = Self::env().caller();
+            let this = self.env().account_id();
+            self.priviledges.insert((caller, Role::Admin(this)), &());
+            self.priviledges.insert((caller, Role::Owner(this)), &());
+        }
+
+        /// Gives a role to an account
+        ///
+        /// Can only be called by an admin role on this contract
+        #[ink(message, selector = 1)]
+        pub fn grant_role(&mut self, account: AccountId, role: Role) -> Result<()> {
+            let key = (account, role);
+            if !self.priviledges.contains(key) {
+                let caller = self.env().caller();
+                let this = self.env().account_id();
+                self.check_role(caller, Role::Admin(this))?;
+                self.priviledges.insert(key, &());
+
+                let event = Event::RoleGranted(RoleGranted {
+                    by: caller,
+                    to: account,
+                    role,
+                });
+                Self::emit_event(self.env(), event);
+            }
+
+            Ok(())
+        }
+
+        /// Revokes a role from an account
+        ///
+        /// Can only be called by an admin role on this contract
+        #[ink(message, selector = 2)]
+        pub fn revoke_role(&mut self, account: AccountId, role: Role) -> Result<()> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+            self.check_role(caller, Role::Admin(this))?;
+            self.priviledges.remove((account, role));
+
+            let event = Event::RoleRevoked(RoleRevoked {
+                by: caller,
+                from: account,
+                role,
+            });
+            Self::emit_event(self.env(), event);
+
+            Ok(())
+        }
+
+        /// Returns true if account has a role
+        #[ink(message, selector = 3)]
+        pub fn has_role(&self, account: AccountId, role: Role) -> bool {
+            self.priviledges.get((account, role)).is_some()
+        }
+
+        /// Terminates the contract.
+        ///
+        /// can only be called by the contract owner
+        #[ink(message, selector = 4)]
+        pub fn terminate(&mut self) -> Result<()> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+            self.check_role(caller, Role::Owner(this))?;
+            self.env().terminate_contract(caller)
+        }
+
+        fn check_role(&self, account: AccountId, role: Role) -> Result<()> {
```

Comment:
Looks like this would be a better candidate for the external call (`#[ink(message]`) than `has_role`, doesn't it?
In the end most messages return some kind of `Result`...

Reply:
> 1. I find the setup with bytecode replacement incredibly hacky and kinda scary :/ There's also a strong "teach by example" factor here - if we release our first batch of contracts with a dirty trick like that, people will start using it as something completely normal and that looks like asking for trouble (sooner or later some replacement will glitch). Since everything happens within a workflow script that is also responsible for compilation, what would you think about an alternative in which the access control address is inserted into the rust source instead of into compiled bytecode? The effect should be more or less the same, but I find such a solution much safer and more elegant.

I actually don't find that all that hacky - I've seen it done for Solidity contacts as gas optimization. Conversely I'd say inserting into source code is more error prone - compilation to WASM is non-deterministic, on top of that it complicates pipeline - one would need to compile the same contract thrice, once per environment.

> 2. _(maybe a bit outside of the scope)_ Would it be useful for regular users to be able to interact with access contol contact? Something like: if I'm an Admin of some contract X, I should be able to grant/revoke `Admin(X)`/`Owner(X)` Role over this contract to whoever I feel like?

it's already possible - grant_role, revoke_role